oidc/provider.go

```diff
@@ -516,19 +515,25 @@ func (s *ProviderConfigSyncer) Run() chan struct{} {
 			select {
 			case <-s.clock.After(next.after()):
 				next = next.step(s.sync)
-			case <-stop:
+			case <-ctx.Done():
+				s.initialSyncComplete()
 				return
 			}
 		}
 	}()
-
-	return stop
 }
 
 func (s *ProviderConfigSyncer) WaitUntilInitialSync() {
 	s.initialSyncWait.Wait()
 }
 
+func (s *ProviderConfigSyncer) initialSyncComplete() {
+	if !s.initialSyncDone {
+		s.initialSyncWait.Done()
```

Comment:
I like to keep the `Done()` call as close to the `Add(1)` call as possible so that it's easier to reason about.

Can we `defer s.initialSyncWait.Done()` in the anonymous goroutine in `Run` instead?


Alternatively, if the initial sync only ever completes once, consider using a `sync.Once` instead.

Reply:
I don't think we can use a `defer` since the function runs as a loop and it wont get executed till the function returns.

I also don't think we can add the check at the end of each loop anyways.  The retry logic will result in `next` being updated without the sync completing.

Let me know if I am missing something, thank you!

Reply:
Ok, the closer I look at this, I don't like the usage of wait group. As the name implies, wait group is best for waiting on a _group_ of operations to complete, but we're only ever waiting on a single thing here.

What if instead of using a waitgroup we used a channel. When the initial sync completes, we close the channel.

`WaitUntilInitialSync()` would just read from the channel. If the initial sync is still in progress, it will correctly wait until the channel is closed. If the initial sync is already completed, it will return immediately.

Reply:
I just pushed an update to switch the `waitGroup` to a channel.  Let me know if you notice anything else, thank you!